File: tests/typesafety/python/entity_ids.py
# expect: passes
# title: Shop and supplier accounts become different types
# improves: transfer_arguments
from typing import NewType

ShopAccount = NewType("ShopAccount", str)
SupplierAccount = NewType("SupplierAccount", str)


def refund(source: ShopAccount, target: SupplierAccount, amount_cents: int) -> str:
    return f"move {amount_cents} from {source} to {target}"


def refund_supplier(shop: ShopAccount, supplier: SupplierAccount) -> str:
    return refund(shop, supplier, 4_500)

File: tests/typesafety/typescript/entity_ids.ts
// expect: passes

declare const shopBrand: unique symbol;
declare const supplierBrand: unique symbol;

type ShopAccount = string & { readonly [shopBrand]: true };
type SupplierAccount = string & { readonly [supplierBrand]: true };

function refund(source: ShopAccount, target: SupplierAccount, amountCents: number): string {
  return `move ${amountCents} from ${source} to ${target}`;
}

export function refundSupplier(shop: ShopAccount, supplier: SupplierAccount): string {
  return refund(shop, supplier, 4_500);
}
